### PR TITLE
Add L1 PulseCapture

### LINKED ISSUE #505: LPC40xx: Pulse Capture

### PR DESCRIPTION
**Pull request by charlesmacd**
Resolve #505

### CODE REVIEW

File: library/L1_Peripheral/lpc40xx/capture.hpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        units::frequency::hertz_t frequency,
+        IsrPointer isr = nullptr, int32_t priority = -1) const override
+    {
+        // NOTE: Same initialization as Timer library's Initialize()
+        system_controller_.PowerUpPeripheral(timer_.channel.power_id);
+        SJ2_ASSERT_FATAL(
+            frequency != 0_Hz,
+            "Cannot have zero ticks per microsecond, please choose 1 or more.");
+
+        // Set Prescale register for Prescale Counter to milliseconds
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/capture.hpp]: Isn't this just creating a divider. I'm not sure what you mean by milliseconds.

File: library/L1_Peripheral/lpc40xx/capture.hpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+                .interrupt_service_routine = timer_.isr,
+                .enable_interrupt = true,
+                .priority = priority
+        });
+
+        // Disable both capture channels
+        ConfigureCapture(kCaptureChannel0, kCaptureEdgeModeNone);
+        ConfigureCapture(kCaptureChannel1, kCaptureEdgeModeNone);
+        EnableCaptureInterrupt(kCaptureChannel0, false);
+        EnableCaptureInterrupt(kCaptureChannel1, false);
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/capture.hpp]: Should you do this first before you register the ISR. If these are defined as true before you get to this point, there is a chance that an event can occur between the registration of the ISR and this disabling of the ISR. This is more evident in a threaded system where the task controlling this peripheral is of low priority and not getting as much CPU time.

File: library/L1_Peripheral/lpc40xx/capture.hpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        else
+        {
+            timer_.channel.capture_pin0.SetPinFunction(3);
+        }
+
+        static constexpr bit::Mask kModeBits[2] =
+        {
+            bit::CreateMaskFromRange(0, 1),
+            bit::CreateMaskFromRange(3, 4)
+        };
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/capture.hpp]: Something is also doable with bit::Masks is that they are actually just a structure with a uint8_t for position and width. Rather than making an array of these you could simply do this:

```suggestion
        static constexpr bit::Mask kModeBits =
        {
            .position =  3 * which,
            .width = 2,
        };
```
This will eliminate some of the code below. Be sure to comment why this is being done though.

File: library/L1_Peripheral/lpc40xx/capture.hpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+        channel.timer_register->IR = bit::Insert(channel.timer_register->IR,
+            kResetCaptureInterrupts, kCaptureInterruptFlagBit);
+    }
+
+    struct CaptureStatus_t
+    {
+        uint32_t count;
+        uint32_t flags;
+    };
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/capture.hpp]: Move this to the top with the other structures.

File: library/L1_Peripheral/lpc40xx/capture.hpp [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        {
+            bit::CreateMaskFromRange(2),
+            bit::CreateMaskFromRange(5)
+        };
+
+        int which = 0;
+        if (channel == kCaptureChannel1)
+        {
+            which = 1;
+        }
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/capture.hpp]: If you set kCaptureChannel1 ==> 1 and kCaptureChannel0 == 0, then you can do this and omit the comparison.

```suggestion
        which = static_cast<int>(capture_channel_)
```

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+void DummyCallback(void) {}
+
+TEST_CASE("Testing lpc40xx Capture", "[lpc40xx-capture]")
+{
+    LPC_TIM_TypeDef test_timer_register;
+
+    constexpr Capture::CaptureChannelNumber kChannelList[] = {
+      Capture::kCaptureChannel0,
+      Capture::kCaptureChannel1,
+    };
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: make sure that `make format-code` hits this file as well.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        .capture_pin1 = capture1_input_pin
+    };
+
+    const Capture::CaptureChannel_t kTestTimer =
+    {
+        .channel = kTestTimerPartial,
+        .isr = sjsu::lpc40xx::Capture::TimerHandler<kTestTimerPartial>
+    };
+
+    memset(&test_timer_register, 0, sizeof(test_timer_register));
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Can you put this right below the creation of the `test_timer_register`. Makes it easier to see how the registers was configured before it was assigned to something. For example, its referenced in the kTestPartial, then cleared, which may give someone the idea, that if the tests are failing its due to the fact that you did not clear your register data, when you actually did, but they missed it because they read the code top to bottom, then found what they believed to be the actual cause of the problem, when in actuality, this part is fine, and something they did broke the code.

Basically, could result in a red herring.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+#include "utility/enum.hpp"
+#include "utility/status.hpp"
+#include "utility/units.hpp"
+
+
+namespace sjsu::lpc40xx
+{
+EMIT_ALL_METHODS(Capture);
+
+void DummyCallback(void) {}
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Could you either put this in a anonymous namespace so the link name does not class against other tests in the `sjsu::lpc40xx` (think about how many other tests probably have a function for an ISR named DummyCallback). The other option is to create a non-capturing lambda in the TEST_CASE.
```suggestion
namespace {
void DummyCallback(void) {}
}
```

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          position = 3;
+        }
+
+        for (size_t mode = 0; mode < std::size(kModeList); mode++)
+        {
+          test_subject.ConfigureCapture(kChannelList[channel], kModeList[mode]);
+          CHECK(bit::Extract(test_timer_register.CCR, position, 2) ==
+                kModeList[mode]);
+        }
+        }
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Incorrect indentation for the curly brace. Please use `make format-code` to fix this.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+    Capture test_subject(kTestTimer,
+        mock_system_controller.get(),
+        mock_interrupt_controller.get());
+
+    SECTION("Capture Initialize")
+    {
+        constexpr units::frequency::hertz_t kTestFrequency = 1_MHz;
+
+        test_subject.Initialize(kTestFrequency, &DummyCallback);
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Missing a test for interrupt.RegisterIsr().

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        test_subject.EnableCaptureInterrupt(kChannelList[channel], false);
+        CHECK(bit::Read(test_timer_register.CCR, position) == false);
+        }
+    }  // end section capture interrupt
+
+    SECTION("Capture Interrupt Handler")
+    {
+      test_timer_register.IR = 0;
+      kTestTimer.isr();
+      CHECK(bit::Extract(test_timer_register.IR, 4, 2) == 0b11);
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Missing a test for a call to DummyCallback, make sure that your ISR does in fact call it. It may be helpful to use the FFF library to mock the calllback and get invocation history.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    }  // end section capture interrupt handler
+}  // end test case
+}  // namespace sjsu::lpc40xx
+
+
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Too many newline characters here.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
+#include <cstdint>
+
+#include "L1_Peripheral/cortex/interrupt.hpp"
+#include "L1_Peripheral/interrupt.hpp"
+#include "L0_Platform/lpc40xx/LPC40xx.h"
+#include "L1_Peripheral/lpc40xx/timer.hpp"
+#include "L1_Peripheral/lpc40xx/capture.hpp"
+#include "L4_Testing/testing_frameworks.hpp"
+#include "utility/enum.hpp"
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: I don't think you used this at all in this file.

File: library/L1_Peripheral/lpc40xx/test/capture_test.cpp [hunk 14]
```diff
@@ -0,0 +1,153 @@
+#include <cstdint>
+
+#include "L1_Peripheral/cortex/interrupt.hpp"
+#include "L1_Peripheral/interrupt.hpp"
+#include "L0_Platform/lpc40xx/LPC40xx.h"
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/test/capture_test.cpp]: Put this at the top as 0 is higher in the alphabetical order.

File: demos/sjtwo/pulse_capture/source/main.cpp [hunk 15]
```diff
@@ -0,0 +1,70 @@
+#include <cstdint>
+
+#include "L1_Peripheral/lpc40xx/timer.hpp"
```
Comment (kammce) [inline demos/sjtwo/pulse_capture/source/main.cpp]: I don't think you need this here.

File: library/L1_Peripheral/lpc40xx/pulse_capture.hpp [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+
+  /// List of available channels per timer to capture from
+  enum CaptureChannelNumber : uint8_t
+  {
+    kChannel0,
+    kChannel1
+  };
+
+  /// Structure used to define attributes of a timer capture channel
+
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/pulse_capture.hpp]:
```suggestion
```
Delete extra space between structure and comment.

File: library/L1_Peripheral/lpc40xx/pulse_capture.hpp [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+        {
+          status.count = channel.timer_register->CR0;
+        }
+        else
+        {
+          status.count = channel.timer_register->CR1;
+        }
+
+        (*channel.user_callback)(status);
+        }
```
Comment (kammce) [inline library/L1_Peripheral/lpc40xx/pulse_capture.hpp]:
```suggestion
        }
```
Got a bit confused here and thought you were not clearing the interrupt flags here. Can you move this curly brace into the correct place.

File: library/L1_Peripheral/pulse_capture.hpp [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+                            int32_t interrupt_priority = -1) const = 0;
+
+  /// Select edge type to capture on
+  virtual void ConfigureCapture(CaptureEdgeMode mode) const = 0;
+
+  /// Enable or disable capture interrupts
+  virtual void EnableCaptureInterrupt(bool enabled) const = 0;
+};
+
+} /* namespace sjsu */
```
Comment (kammce) [inline library/L1_Peripheral/pulse_capture.hpp]:
```suggestion
}  // namespace sjsu::lpc40xx
```
Can you change this to the following suggestion like the rest of your code.
